### PR TITLE
Allow users to specify data types for a subset of columns in `read_csv` 

### LINKED ISSUE #10254: [BUG] read_csv fails when not all columns specified in the `dtype` argument
```python
import cudf
import pandas as pd

# generate a CSV:
df = cudf.DataFrame(
    {
        "a": [1, 2, 3],
        "b": ["x", "y", "z"],
        "c": cudf.Series([1, 2, 3], dtype="datetime64[ns]"),
    }
)
df.to_csv("test.csv", index=False)

# this works:
pd.read_csv("test.csv", dtype={"a": "int"})

# this doesn't:
cudf.read_csv("test.csv", dtype={"a": "int"})
```

```python
...
RuntimeError: cuDF failure at: /home/ashwin/workspace/rapids/cudf/cpp/src/io/csv/reader_impl.cu:493: Must specify data types for all active columns
```

### PR DESCRIPTION
**Pull request by vuule**
Fixes #10254

CSV reader previously assumed that all data types are specified by the user, or none. 
This PR changes the logic so that user can pass a map/dictionary to specify type for any subset of columns, and reader infers the type for the remaining columns.
When passing columns as an array, users still need to specify all columns' types, because the array become ambiguous when reading a subset of columns in the file.

### CODE REVIEW

File: cpp/src/io/csv/reader_impl.cu [hunk 1]
```diff
@@ -622,6 +617,49 @@ std::vector<column_buffer> decode_data(parse_options const& parse_opts,
...
+                     column_types,
+                     stream);
+
+  // compact column_types to only include active columns
+  std::vector<data_type> active_col_types;
+  std::copy_if(column_types.cbegin(),
+               column_types.cend(),
+               std::back_inserter(active_col_types),
+               [&column_flags, &column_types](auto const& dtype) {
+                 auto const idx = &dtype - column_types.data();
```
Comment (bdice) [inline cpp/src/io/csv/reader_impl.cu:656]: Avoid raw pointer arithmetic here.
```suggestion
                 auto const idx = std::distance(column_types.data(), &dtype);
```

**Review by bdice — APPROVED** [review]
Looks great overall! Only one minor change request.

File: cpp/src/io/csv/reader_impl.cu [hunk 2]
```diff
@@ -457,116 +457,111 @@ std::pair<rmm::device_uvector<char>, selected_rows_offsets> select_data_and_row_
...
+    } else if (stats.datetime_count > 0L) {
+      column_types[col_idx] = timestamp_type.id() == cudf::type_id::EMPTY
+                                ? data_type(cudf::type_id::TIMESTAMP_NANOSECONDS)
+                                : timestamp_type;
+    } else if (stats.bool_count > 0L) {
+      column_types[col_idx] = data_type(cudf::type_id::BOOL8);
+    } else if (stats.float_count > 0L ||
+               (stats.float_count == 0L && int_count_total > 0L && stats.null_count > 0L)) {
+      // The second condition has been added to conform to
+      // PANDAS which states that a column of integers with
```
Comment (bdice) [inline cpp/src/io/csv/reader_impl.cu:553]: The library name is usually written in lowercase.
```suggestion
      // pandas which states that a column of integers with
```

**Comment by galipremsagar** [comment]
is this intended for `22.04` or `22.06`? - Asking this because the PR and issue are on two different boards.

**Comment by shwina** [comment]
Apologies for missing the pings about this one. 22.06 should be fine.

**Review by ttnghia — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**